Refactored reconciler checks

TylerRHoyer opened the pull request:
Closes #110 .

Refactored reconciler checks to be more efficient as outlined in #110.

Checklist before submitting:
* [ ] Added entry to `CHANGELOG.md`
* [x] Added/updated relevant tests
* [x] Added/updated documentation

LPGhatguy commented on lib/Reconciler.lua:
```diff
@@ -64,75 +56,47 @@ local Reconciler = {}
 
-	if isPrimitiveElement(element) then
-		-- We're destroying a Roblox Instance-based object
+	if type(element) == "table" then
```
Instead of indenting the entire function one level, we should probably just early-out with an error if `element` isn't a table.

That being said, in this function, `element` should _always_ be a table, so this check is just redundant.

LPGhatguy commented on lib/Reconciler.lua:
```diff
@@ -156,115 +120,118 @@ end
 function Reconciler._mountInternal(element, parent, key, context)
-	if isPrimitiveElement(element) then
-		-- Primitive elements are backed directly by Roblox Instances.
+	if type(element) == "table" then
```
Same here, probably not worth checking if the element is a table; we can do typechecking assertions in `mount` though in a later PR, since it's a user-facing API.

TylerRHoyer commented:
I've removed some checks, and moved others to the public version of the method. Also, changed the logic for the logic in the public version of the reconcile method. It was checking if the instanceHandle was nil then indexing it. It would throw a nasty error if instanceHandle wasn't a table, so I made it first check if the instanceHandle is a table then check if it has the isInstanceHandle symbol.

LPGhatguy commented:
If possible, I'd like to keep the scope of PRs like this pretty small.

Can we scale this one back to _just_ refactoring the checks on `element.component`? That's a change I would be cool with merging sooner, but reviewing some of the other things here is more difficult.

TylerRHoyer commented:
I moved error in unmounts back into the elseif statement. Unfortunately refactoring the `element.component` checks means also moving the `type(element)` checks, and if we move them then the best spot to put them is out of the internal functions, as you mentioned. I'm not sure what else could be removed from this commit, do you see anything?

LPGhatguy commented on lib/Reconciler.lua:
```diff
@@ -140,6 +100,22 @@ end
+			-- Ignore booleans of either value
+			-- See https://github.com/Roblox/roact/issues/14
+			return nil
+		end
```
Pretty sure moving this check here will break lots of code, since everything in the reconciler uses `_mountInternal`, not `mount.`

LPGhatguy commented on lib/Reconciler.lua:
```diff
@@ -264,10 +241,6 @@ function Reconciler._mountInternal(element, parent, key, context)
-	elseif typeof(element) == "boolean" then
-		-- Ignore booleans of either value
-		-- See https://github.com/Roblox/roact/issues/14
-		return nil
```
We should keep this check inside this function. If we're going to move it, let's put it first so that we can avoid indexing into a bool/nil element.

LPGhatguy commented:
This has given me some good ideas for how I think we want to check for components in the future.

I'm afraid that just flat-out removing the current `isFoo` checks causes the code to be a little less readable.

I started up a new branch named `element-kind-checks` that will have a PR open soon that takes a slightly different approach. Instead of having a bunch of `isWhatever` checks, I added a `getElementKind` function that classifies an element in one go.

LPGhatguy commented:
I think I'm going to close this in favor of #123. Thanks for the work on this, it was super useful in figuring out where to take this sort of code I think.
